Dev gradient

EE-LiuYunhao opened the pull request:
# Gradient Backpropagation

Optimize the bridge between the robots' forward kinematics and the primitives' forward kinematics, and modify the gradient backpropagation accordingly. 

## Detailed modifications

### `plb/engine/controller.py`
A newly-added abstraction to generalize the common behaviours of `plb.engine.RobotController` and `plb.engine.primitive.Primitives` _(the controller for free primitives)_. The controllers are used in the MPM simulator. The following interfaces are exposed to the simulator:
1. `controller.set_action(index, substeps, action)` -> set action for whatever being controlled by the controller
1. `controller.forward_kinematics(index)` -> execute the previously set actions
1. `controller.forward_kinematics.grad(index)` -> gradient backpropagation of the `forward_kinematics` method

The controller's forward kinematics and its gradient backpropagation method are designed in the above way to be compatible with Taichi's complex kernel. 

### `plb/engine/mpm_simulator.py`
1. Utilize the `Controller` interfaces to control movable stuff to interact with the environment
1. Allows `torch.Tensor`

### `plb/engine/primitive/primitives.py`
Inheriting the `Controller`

### `plb/engine/primitive/primive_base.py`
Add a complex kernel called `apply_robot_forward_kinematics`. 

The kernel is designed for the `RobotController` to directly apply the results from the robot's forward kinematics onto each primitive. Previously, differentiation and conversion were necessary since the robot's FK gives each link's pose, but a primitive only accepts a 6-dim velocity as inputs. The new kernel offers a shortcut to apply poses on primitives without converting them into speeds directly. 

### `plb/engine/taichi_env.py`
Use the `Controller` interfaces to replace the original `Primitives` and `RobotController`. 

### `plb/urdfpy/__init__.py`
The `DEVICE`, i.e.,  where the robots' components are loaded, is exposed. 

### `plb/urdfpy/diff_fk.py`

Gradient backpropagation is fixed here. 

### `plb/urdfpy/manipulation.py`

Use OrderDict to replace Dict for `robot._link_map` to ensure that the backpropagation happens precisely in the forward pass's reversed order.

## Tests

### Unit tests
All passed
![image](https://user-images.githubusercontent.com/43565614/153831802-c2d2a5d7-03e8-4e55-bd4a-50aa8c89a1c6.png)

#### Existing unit tests

Some tests are modified due to the alteration of the method signatures and interfaces. 

#### Newly added test: `tests/unit/test_robot_grad.py`

Ensure the gradient pass of the forward kinematics is correct. 

### Functional tests
N/A, because of the OOM issue of TaichiEnv.

EE-LiuYunhao committed:
Robot's gradients added to
NOT TESTED YEt
To be optimized for the for/back-ward paths

EE-LiuYunhao committed:
Use the URDF primitives for UT

EE-LiuYunhao committed:
Revert the DOF to pass UT

EE-LiuYunhao committed:
refactor some confusing codes

EE-LiuYunhao committed:
Directly apply FK's result to avoid substraction
NOT TESTED YET

EE-LiuYunhao committed:
Use a controller to deal with FK

EE-LiuYunhao committed:
Implementation to retrieve gradients
NOT TESTED
* Unit tests need to be modified accordingly
* Importing statements are adjusted

EE-LiuYunhao committed:
Fix the unit tests
NO NEW TEST INVOLVED
Still, more unit tests needed to ensure the correctness of the backprop

EE-LiuYunhao committed:
Fix the gradient backpropagation
UNIT TESTS PASSED

EE-LiuYunhao committed:
Orderdict to replace dict in urdfpy.robot

EE-LiuYunhao committed:
Switch back to ur5_primitive

EE-LiuYunhao committed:
Simplify the Controller logic

EE-LiuYunhao committed:
Fix the unit tests

shwnyao commented on plb/engine/mpm_simulator.py:
```diff
@@ -67,7 +77,16 @@ def __init__(self, cfg, primitives=()):
         self.primitives = primitives
+        self.rc = robots_controller
+        """Robots Controller"""
+        self.fpc = free_primitives_controller if free_primitives_controller != None \
```
fpc 是指原來的 primitive嗎 看起來可以同時有多種controller?

shwnyao commented on plb/engine/mpm_simulator.py:
```diff
@@ -319,8 +338,8 @@ def substep_grad(self, s):
-        for i in range(self.n_primitive-1, -1, -1):
-            self.primitives[i].forward_kinematics.grad(s)
+        self.fpc.forward_kinematics.grad(s)
+        self.rc.forward_kinematics.grad(s)
```
fpc和rc之間是不能用polymorphism 去簡化的嗎？

shwnyao commented on plb/engine/mpm_simulator.py:
```diff
@@ -441,8 +460,10 @@ def step(self, is_copy, action=None):
         if action is not None:
-            self.primitives.set_action(
-                start//self.substeps, self.substeps, action)
+            self.fpc.set_action(start//self.substeps, self.substeps, 
```
@EE-LiuYunhao 這一段會導致沒有robot時action輸出None #25

EE-LiuYunhao replied:
Okay will check and resolve it.

EE-LiuYunhao committed:
Add a not_empty property for cases having no robot

This should resolve the bug that torch.cat is concatenating primitives'
values with NONE

EE-LiuYunhao commented:
@shwnyao Robot在act/act grad里面返回None然后被concatenate的问题应该解决了。你检查一下？

shwnyao commented:
@EE-LiuYunhao I tried the following command and this happened. It runs nnv2.
`python3 -m plb.algorithms.tune` 
```
Traceback (most recent call last):
  File "/root/shwnyao/PlasticineLab/plb/optimizer/solver_torch_nnv2.py", line 73, in train
    obs, reward, done, loss_info = self.env.step(action)
  File "/usr/local/lib/python3.6/dist-packages/gym/wrappers/time_limit.py", line 16, in step
    observation, reward, done, info = self.env.step(action)
  File "/root/shwnyao/PlasticineLab/plb/envs/env.py", line 44, in step
    self.taichi_env.step(action)
  File "/root/shwnyao/PlasticineLab/plb/engine/taichi_env.py", line 108, in step
    self.simulator.step(is_copy=self._is_copy, action=action)
  File "/root/shwnyao/PlasticineLab/plb/engine/mpm_simulator.py", line 469, in step
    self.substep(s)
  File "/usr/local/lib/python3.6/dist-packages/taichi/lang/__init__.py", line 861, in decorated
    func(*args, **kwargs)
  File "/root/shwnyao/PlasticineLab/plb/engine/mpm_simulator.py", line 321, in substep
    self.fpc.forward_kinematics(s)
AttributeError: 'Primitives' object has no attribute 'forward_kinematics'
```

EE-LiuYunhao commented:
@shwnyao This is my fault. I will fix it immediately.

EE-LiuYunhao committed:
Solve the bug:
primitives has no attribute named "forward_kinematics"

EE-LiuYunhao commented:
@shwnyao The bug should be solved now. Could you please verify it?

shwnyao committed:
fix(primitives): minor fix of the member variable usage

shwnyao commented:
#25 has been solved.
LGTM
